### PR TITLE
New Package: DependencyExtractionWebpackPlugin

### PR DESCRIPTION
**Pull request by psealock**
Partially Fixes https://github.com/woocommerce/woocommerce-admin/issues/3445

As `@woocommerce` packages continue to grow in number and usefulness, plugins wishing to consume them will continually need to amend webpack's [DependencyExtractionWebpackPlugin](https://github.com/WordPress/gutenberg/blob/master/packages/dependency-extraction-webpack-plugin/README.md) options `RequestToHandle` and `requestToExternal` so that dependencies can be handled correctly.

This PR introduces a woo-flavored DependencyExtractionWebpackPlugin that extends the WP version such that WC packages are also mapped automatically.

### Detailed test instructions:

The best way to test this is to create a new extension using `npm run create-wc-extension`. A project is scaffolded out, including use of DependencyExtractionWebpackPlugin. Once this simple case is confirmed, we can start using this package in wc-admin itself.

1. `npm run build`. This builds the packages for consumption later when we mock our new package (See step 3).
1. `npm run create-wc-extension` and give your new plugin a name.
2. Change this line in `<your-new-plugin>/webpack.config.js` to point to the new package since it does not yet exist in npm. There's probably a nice `npm link` solution here, but my tinkering didn't reveal it and this is easy enough.

```diff
- const WooCommerceDependencyExtractionWebpackPlugin = require( '@woocommerce/dependency-extraction-webpack-plugin' );
+ const WooCommerceDependencyExtractionWebpackPlugin = require("../woocommerce-admin/packages/dependency-extraction-webpack-plugin/src/index");
```
3. `npm start`
4. See `<your-new-plugin>/build/index.asset.php` exist and declare an empty array of dependencies because there aren't any just yet.
5. Install some dependencies: `npm install @wordpress/components` and `npm install @woocommerce/navigation`.
6. Use those dependencies by adding this to `<your-new-plugin>/src/index.js`:

```js
import { Button } from "@wordpress/components";
import { getHistory } from "@woocommerce/navigation";

console.log(Button, getHistory);
```
7. restart the proces `npm start`
8. See `<your-new-plugin>/build/index.asset.php` now reflect two dependencies.
9. Check the built js and see the dependencies gathered from `this`, which is the `window`
9. Bonus: Activate the plugin and visit the homescreen. See our consoled components.
10. Confirm that `window.wp.components` exists (wc-admin doesn't load it).

### Changelog Note:

- Dev: Add WooCommerceDependencyExtractionWebpackPlugin package

### CODE REVIEW

File: bin/starter-pack/_package.json
Comment (nerrad) [inline bin/starter-pack/_package.json:19]: Aren't you wanting to include `@woocomerce/dependency-extraction-webpack-plugin` here?

File: packages/dependency-extraction-webpack-plugin/src/index.js
Comment (nerrad) [inline packages/dependency-extraction-webpack-plugin/src/index.js:32]: This makes the assumption that every import on the `@woocommerce/*` namespace will have an external (and dependency). Woo Blocks uses the `@woocommerce/*` namespace for a number of aliases only. It'd be good if there would be a way we could signal skipping over aliased namespaces via config.

**Review by nerrad — COMMENTED** [review]
Love the idea here Paul!  I think it'll be nice for Woo packages to be able to simplify their Webpack configuration using this.

Just a couple comments for review.

File: bin/starter-pack/_package.json
Comment (psealock) [reply bin/starter-pack/_package.json:19]: Yes! Good eye @nerrad. I left it this way temporarily so that an initial `npm install` for testing this PR will work because that package does not yet exist yet.

**Comment by psealock** [comment]
Thanks for the feedback @nerrad ! This is ready for another review.

>Woo Blocks uses the @woocommerce/* namespace for a number of aliases only

Good point. Instead of putting the onus on consumers to declare aliases, I added a build step to dynamically create a list of packages before packages are built in https://github.com/woocommerce/woocommerce-admin/pull/5198/commits/d0c73a80aeaec1ae70816675b1019ad57b1a7660. The list is used by `wooRequestToExternal` and `wooRequestToHandle` to check if the request should be an external instead of just checking if it begins with `@woocommerce/`.

>Aren't you wanting to include @woocomerce/dependency-extraction-webpack-plugin here?

Yup, this was for easier testing. I'll replace before merge.

File: packages/dependency-extraction-webpack-plugin/src/index.js
Comment (joshuatf) [inline packages/dependency-extraction-webpack-plugin/src/index.js:2]: I may have set things up wrong, but this is causing me some trouble.  Do I need to link something in order for this to work?

Screen Shot 2020-10-05 at 11 27 53 AM

This was after running `npm start` from the newly created extension directory.

**Review by joshuatf — COMMENTED** [review]
This is a great addition (and clean-up) to the WC ecosystem.  Thanks for taking the time to do this, Paul 🥇 

Had a little trouble testing this with the latest commits, but code looks pretty good to me.  @nerrad or @senadir can you confirm this works for your use case?

File: packages/dependency-extraction-webpack-plugin/src/index.js
Comment (psealock) [reply packages/dependency-extraction-webpack-plugin/src/index.js:2]: Thanks for taking this for a spin @joshuatf. The problem lies step 3 of the instructions, which assumed the sibling wc-admin plugin had built packages. I changed the instructions to reflect this if you could give it another shot.

The trouble lies in creating a dependency on a package that isn't published yet, so this will be a lot simpler once the package in on npm.

**Review by joshuatf — APPROVED** [review]
Woo!  Got it up and running and dependencies are being loaded correctly.  LGTM after a rebase to fix up any conflicts.  ✅

**Pull request merged**
